pkg/controllers/leaderworkerset_controller.go

```diff
@@ -537,51 +540,105 @@ func (r *LeaderWorkerSetReconciler) iterateReplicas(ctx context.Context, lws *le
-		return 0, 0, err
+		return nil, err
 	}
 	sortedSts := utils.SortByIndex(func(sts appsv1.StatefulSet) (int, error) {
 		return strconv.Atoi(sts.Labels[leaderworkerset.GroupIndexLabelKey])
 	}, stsList.Items, int(stsReplicas))
 
 	// Once size==1, no worker statefulSets will be created.
 	noWorkerSts := *lws.Spec.LeaderWorkerTemplate.Size == 1
-	processReplica := func(index int32) (ready bool) {
-		nominatedName := fmt.Sprintf("%s-%d", lws.Name, index)
+
+	for idx := int32(0); idx < stsReplicas; idx++ {
+		nominatedName := fmt.Sprintf("%s-%d", lws.Name, idx)
 		// It can happen that the leader pod or the worker statefulset hasn't created yet
 		// or under rebuilding, which also indicates not ready.
-		if nominatedName != sortedPods[index].Name || (!noWorkerSts && nominatedName != sortedSts[index].Name) {
-			return false
+		if nominatedName != sortedPods[idx].Name || (!noWorkerSts && nominatedName != sortedSts[idx].Name) {
+			states[idx] = replicaState{
+				ready:   false,
+				updated: false,
+			}
+			continue
 		}
 
-		podTemplateHash := revisionutils.GetRevisionKey(&sortedPods[index])
-		if !(podTemplateHash == revisionKey && podutils.PodRunningAndReady(sortedPods[index])) {
-			return false
-		}
+		leaderUpdated := revisionutils.GetRevisionKey(&sortedPods[idx]) == revisionKey
+		leaderReady := podutils.PodRunningAndReady(sortedPods[idx])
 
 		if noWorkerSts {
-			return true
+			states[idx] = replicaState{
+				ready:   leaderReady,
+				updated: leaderUpdated,
+			}
+			continue
 		}
 
-		stsTemplateHash := revisionutils.GetRevisionKey(&sortedSts[index])
-		return stsTemplateHash == revisionKey && statefulsetutils.StatefulsetReady(sortedSts[index])
+		workersUpdated := revisionutils.GetRevisionKey(&sortedSts[idx]) == revisionKey
+		workersReady := statefulsetutils.StatefulsetReady(sortedSts[idx])
+
+		states[idx] = replicaState{
+			ready:   leaderReady && workersReady,
+			updated: leaderUpdated && workersUpdated,
+		}
 	}
 
-	var skip bool
-	var continuousReadyReplicas, lwsUnreadyReplicas int32
+	return states, nil
+}
+
+func rollingUpdatePartition(states []replicaState, stsReplicas int32, rollingStep int32, currentPartition int32) int32 {
+	continuousReadyReplicas := calculateContinuousReadyReplicas(states)
 
-	for index := stsReplicas - 1; index >= 0; index-- {
-		replicaReady := processReplica(index)
-		skip = skip || !replicaReady
+	// Update up to rollingStep replicas at once.
+	var minAllowedPartition = utils.NonZeroValue(stsReplicas - continuousReadyReplicas - rollingStep)
 
-		if replicaReady && !skip {
-			continuousReadyReplicas++
+	// Replicas that are not ready or expected to be updated (since idx >= partition) given current partition.
+	// Used to determine how much the partition can be reduced.
+	var unavailable int32
+	for idx := 0; idx < int(stsReplicas-continuousReadyReplicas); idx++ {
+		if !states[idx].ready || idx >= int(currentPartition) {
+			unavailable++
```

Comment:
I guess the `unavailable` is not right here because `idx >= partition` doesn't mean it's unavailable?